### PR TITLE
Update musl to v1.2.2

### PR DESCRIPTION
**Pull request by sbc100**
The changes here mostly come from running the update_musl.py script         
against the following commit:                                            
                                                                         
https://github.com/emscripten-core/musl/tree/merge_v1.2.2                
                                                                         
Notes:                                                                   
                                                                         
- `weak` is defined in musl-internal headers so some of our usage of        
  `__attribute__((weak))` was switched to `__weak__` (which is a         
  synonym).                                                              
                                                                         
- Quite a few changes needed in library_pthread.js due to changes to        
  the internal representation of __pthread struct in C.                  
  - removal of emscripten-specific threadStatus field.  This can         
    can be fully replaced with the updated `detach_state` that           
    was added to musl (along with enum values for states!) along         
    with the existing `cancel` bit.                                      
  - `attr` is no longer stored on `__pthread`; it was only duplicating   
    other fields anyway.  We now include `pthread_getattr_np.c` which                                               
    can pull those field out on demand into at attr structure.       

Fixes:  #8740

### CODE REVIEW

File: src/library_pthread.js [hunk 1]
```diff
@@ -59,8 +59,13 @@ var LibraryPThread = {
       // Allocate memory for thread-local storage.
       var tlsMemory = _malloc({{{ cDefine('PTHREAD_KEYS_MAX') * 4 }}});
       for (var i = 0; i < {{{ cDefine('PTHREAD_KEYS_MAX') }}}; ++i) HEAPU32[tlsMemory/4+i] = 0;
+<<<<<<< HEAD
       Atomics.store(HEAPU32, (tb + {{{ C_STRUCTS.pthread.tsd }}} ) >> 2, tlsMemory); // Init thread-local-storage memory array.
```
Comment (vadimkantorov) [inline src/library_pthread.js:63]: seems a conflict...

File: system/lib/libc/crt1.c [hunk 2]
```diff
@@ -12,7 +12,7 @@
 #include <stdlib.h>
 #include <wasi/api.h>
 
-__attribute__((weak)) void __wasm_call_ctors(void);
+__attribute__((__weak__)) void __wasm_call_ctors(void);
```
Comment (kripken) [inline system/lib/libc/crt1.c:15]: Is there a reason to prefer the underscores? I've never seen this form I don't think.

File: system/lib/libc/musl/src/locale/newlocale.c [hunk 3]
```diff
@@ -1,47 +1,71 @@
...
+	if (!memcmp(&tmp, &default_ctype_locale, sizeof tmp))
+		return &default_ctype_locale;
 
-	if (!j)
-		return C_LOCALE;
-	if (j==1 && tmp.cat[LC_CTYPE]==&__c_dot_utf8)
-		return UTF8_LOCALE;
+	/* If no builtin locale matched, attempt to allocate and copy. */
+	/* XXX EMSCRIPTEN .. */
+	/*if ((loc = malloc(sizeof *loc))) *loc = tmp;*/
```
Comment (kripken) [inline system/lib/libc/musl/src/locale/newlocale.c:58]: Are these three lines a difference from upstream? (Then `ifdef` could be used?) And, why the divergence?

File: system/lib/libc/musl/src/thread/pthread_mutex_timedlock.c [hunk 4]
```diff
@@ -1,24 +1,86 @@
...
-	int r, t, priv = (m->_m_type & 128) ^ 128;
+	int type = m->_m_type;
+	int r, t, priv = (type & 128) ^ 128;
 
-	r = pthread_mutex_trylock(m);
+	r = __pthread_mutex_trylock(m);
 	if (r != EBUSY) return r;
+
+#ifndef __EMSCRIPTEN__
+	if (type&8) return pthread_mutex_timedlock_pi(m, at);
```
Comment (kripken) [inline system/lib/libc/musl/src/thread/pthread_mutex_timedlock.c:71]: What is the `(type&8)` stuff that is ignored? Something new musl added that we don't need? (Could use a comment somewhere perhaps.)

File: system/lib/update_musl.py [hunk 5]
```diff
@@ -55,6 +54,11 @@ def main():
   # Copy new version into place
   shutil.copytree(musl_dir, local_src, ignore=ignore)
 
+  # Create versoin.h
```
Comment (kripken) [inline system/lib/update_musl.py:57]:
```suggestion
  # Create version.h
```

File: tests/core/test_float_builtins.out [hunk 6]
```diff
@@ -2,7 +2,7 @@
 0.567800 : 1.010100 : 21.320100
 0.123400 : 0.101010 : 12.012300
 0.123400 : 0.101010 : 12.012300
-0.304824 : 1.001016 : 104168858480774546980864.000000
+0.304824 : 1.001016 : 104168858480774530203648.000000
```
Comment (kripken) [inline tests/core/test_float_builtins.out:5]: Why did this change?

File: tests/test_core.py [hunk 7]
```diff
@@ -8469,7 +8469,7 @@ def setUp(self):
 # Main wasm test modes
 wasm0 = make_run('wasm0', emcc_args=['-O0'])
 wasm0g = make_run('wasm0g', emcc_args=['-O0', '-g'])
-wasm1 = make_run('wasm1', emcc_args=['-O1'])
+wasm1 = make_run('wasm1', emcc_args=['-O1', '-g'])
```
Comment (kripken) [inline tests/test_core.py:8472]:
```suggestion
wasm1 = make_run('wasm1', emcc_args=['-O1'])
wasm1g = make_run('wasm1', emcc_args=['-O1', '-g'])
```
(I assume the intent was to add a variant with g, not replace the existing one?)

**Review by kripken — COMMENTED** [review]
Great work!

Please update the changelog and the readme.

What is the cause of the code size regressions?

File: system/lib/libc/crt1.c (hunk 2 above)
Comment (sbc100) [reply system/lib/libc/crt1.c:15]: Sorry I should have written more notes in the PR description.

This was needed because musl internally defines `weak` .. so for any code that gets compiled with musl itself we can't use the word `weak` anywhere.

File: tests/test_core.py (hunk 7 above)
Comment (sbc100) [reply tests/test_core.py:8472]: Reverted.

File: tests/core/test_float_builtins.out (hunk 6 above)
Comment (sbc100) [reply tests/core/test_float_builtins.out:5]: Good question... do you have any idea?   Do you know how to tell which one is correct?   I didn't look into it much yet.

File: system/lib/libc/musl/src/thread/pthread_mutex_timedlock.c (hunk 4 above)
Comment (sbc100) [reply system/lib/libc/musl/src/thread/pthread_mutex_timedlock.c:71]: Good question, I'm not sure exactly.  I will investigate more.  Looks like special internal lock flag that uses FUTEX_LOCK_PI under the hood.

Comment (kleisauke) [reply system/lib/libc/musl/src/thread/pthread_mutex_timedlock.c:71]: This just reverts commit http://git.musl-libc.org/cgit/musl/commit/?id=54ca677983d47529bab8752315ac1a2b49888870, I think. Perhaps adding
```
// XXX Emscripten PI mutexes are unsupported (reverts musl commit 54ca677983d47529bab8752315ac1a2b49888870)
```
would help?

File: tests/core/test_float_builtins.out (hunk 6 above)
Comment (kripken) [reply tests/core/test_float_builtins.out:5]: I'm not sure. Running the testcase in gcc I get a slightly different result from both of these, which suggests to me that there isn't an absolute precision that libc is required to provide. Or that if there is, this doesn't regress us any further at least. Also, there are no emscripten changes to `pow.c`, so this is likely either an innocuous musl change, or an improvement. So seems good to me.

Comment (kleisauke) [reply tests/core/test_float_builtins.out:5]: fwiw, on Alpine Linux (which uses musl too) this is also different.

<details>
  <summary>Details</summary>

```bash
$ docker run -v $(pwd):/app -w /app -it alpine:latest \
   apk add build-base && \
   gcc tests/core/test_float_builtins.c -lm && \
   ./a.out > tests/core/test_float_builtins.out
$ git diff HEAD tests/core/test_float_builtins.out
```

```diff
diff --git a/tests/core/test_float_builtins.out b/tests/core/test_float_builtins.out
index 987a82b9d..2cfbde14e 100644
--- a/tests/core/test_float_builtins.out
+++ b/tests/core/test_float_builtins.out
@@ -2,8 +2,8 @@
 0.567800 : 1.010100 : 21.320100
 0.123400 : 0.101010 : 12.012300
 0.123400 : 0.101010 : 12.012300
-0.304824 : 1.001016 : 104168858480774530203648.000000
-0.000000 : 2.421390 : 101648962858951235671184182497867417669061891127967232769386628692937206561802464554722378383360.000000
+0.304824 : 1.001016 : 104168858480774568304640.000000
+0.000000 : 2.421390 : 101648962858951233521793526030185540119025481904197943243062538336984236579371061407834909442048.000000
 inf : 3.091515 : 3.091515 : 1
 sqrt(x) = 4.000000
 ceil(x) = 16.000000
```
</details>

File: system/lib/libc/musl/src/thread/pthread_mutex_timedlock.c (hunk 4 above)
Comment (sbc100) [reply system/lib/libc/musl/src/thread/pthread_mutex_timedlock.c:71]: Done

File: tests/core/test_float_builtins.out (hunk 6 above)
Comment (sbc100) [reply tests/core/test_float_builtins.out:5]: I get the same results as your alpine linux test on my desktop too, using both gcc and clang.  However if I switch to -O2 then gcc changes but clang doesn't (for the final line).   

Given that our results are *already* different to the desktop gcc/clang results I think we can let this slide.

File: system/lib/libc/musl/src/locale/newlocale.c (hunk 3 above)
Comment (sbc100) [reply system/lib/libc/musl/src/locale/newlocale.c:58]: Investigating the history of this one line let me to this change: https://github.com/emscripten-core/emscripten/pull/13584

**Comment by sbc100** [comment]
Updated with thread state management that matched more closely the the way musl does it.  We now use musl's `detach_state` for in `__pthread` which now has an enum of different values.

File: system/lib/libc/musl/include/features.h [hunk 8]
```diff
@@ -6,6 +6,7 @@
 #endif
 
 #if defined(_DEFAULT_SOURCE) && !defined(_BSD_SOURCE)
+#error YYY
```
Comment (kleisauke) [inline system/lib/libc/musl/include/features.h:9]: Looks like a debug leftover.

File: tests/core/test_float_builtins.out (hunk 6 above)
Comment (kleisauke) [reply tests/core/test_float_builtins.out:5]: I updated my comment because the test ran on my own system. It seems that Alpine also differs in the `asin(x)` test.

> Given that our results are _already_ different to the desktop gcc/clang results I think we can let this slide.

Indeed, I don't think it's worth investigating this further.

**Comment by sbc100** [comment]
Looking at the size regressions here:

The regressions in `tests/code_size` are all pretty small:

```
test                                    old     new     delta   %delta              
code_size/hello_webgl2_wasm.json        15867   16010   143     0.90%               
code_size/hello_webgl2_wasm2js.json     23658   23864   206     0.87%                              
code_size/hello_webgl_wasm.json         15353   15496   143     0.93%               
code_size/hello_webgl_wasm2js.json      23143   23349   206     0.89%               
code_size/random_printf_wasm.json       12496   12992   496     3.97%               
code_size/random_printf_wasm2js.json    17286   17939   653     3.78%  
````

Looking at `tests/metadce` we see two size regressions.

```
hello_world_O0.size                                     11479   12419   940     8.19%              
minimal_main_Oz_USE_PTHREADS_PROXY_TO_PTHREAD.size      17011   17993   982     5.77%
```

Notice that only the O0 version hello world regressed enough to trigger a failure.  O1 and above did not show the regression.  The differences was as follows:

```
$ twiggy diff /tmp/emscripten_test/a.out.wasm  a.out.wasm 
 Delta Bytes │ Item
─────────────┼──────────────────────────────────────────
        +251 ┊ pop_arg
        +152 ┊ data[1]
        +151 ┊ fmt_fp
        +132 ┊ fflush
        +127 ┊ printf_core
        -109 ┊ __fflush_unlocked
         +57 ┊ __lseek
         +41 ┊ getint
         +34 ┊ import wasi_snapshot_preview1::fd_close
         +33 ┊ import wasi_snapshot_preview1::fd_seek
         +24 ┊ strnlen
         +21 ┊ legalfunc$__wasi_fd_seek
         +15 ┊ __stdio_seek
         +13 ┊ __stdio_close
         -12 ┊ __vfprintf_internal
          +9 ┊ type[7]: (i32, i32, i32, i32, i32) -> i32
          -9 ┊ type[8]: (i32, i32, i32, i32, i32) -> i32
          +8 ┊ type[22]: (i32, i64, i32, i32) -> i32
          +7 ┊ __get_tp
          -7 ┊ __pthread_self
          +1 ┊ ... and 19 more.
       +1039 ┊ Σ [39 Total Rows]
```

None of these look particularly actionable.  __fflush_unlocked was merged into fflush which so they cancel each other out.
 pop_arg fmt_fp and printf_core are all related to incremental changes in `system/lib/libc/musl/src/stdio/vfprintf.c` it seems.  Since these are at O0 only I think we can live with them.

For `test_metadce_minimal_pthreads` we did see a 982 byte (5%) increase which is down to. Sadly twiggy crashes here by the delta is in the code section somwhere:

```
$ bloaty a.out.wasm --  /tmp/emscripten_test/a.out.wasm 
    FILE SIZE        VM SIZE    
 --------------  -------------- 
  +6.2%    +966  [ = ]       0    Code
  +5.5%     +90  [ = ]       0    name
   +14%      +7  [ = ]       0    Function
  +0.1%      +1  [ = ]       0    Data
  +5.7% +1.04Ki  [ = ]       0    TOTAL
```

File: system/lib/libc/musl/src/thread/pthread_key_create.c [hunk 9]
```diff
@@ -7,54 +7,91 @@
...
 {
+	sigset_t set;
+	pthread_t self = __pthread_self(), td=self;
+
+	__block_app_sigs(&set);
+	__pthread_rwlock_wrlock(&key_lock);
+
+	__tl_lock();
+	do { td->tsd[k] = 0;
+    printf("XXX self=%p next=%p\n", self, td->next);
```
Comment (kleisauke) [inline system/lib/libc/musl/src/thread/pthread_key_create.c:62]: Looks like a debug leftover.

File: system/lib/pthread/library_pthread.c [hunk 10]
```diff
@@ -825,17 +825,26 @@ EMSCRIPTEN_KEEPALIVE void* _emscripten_main_thread_futex;
...
+  __main_pthread.detach_state = DT_JOINABLE;
   // pthread struct robust_list head should point to itself.
   __main_pthread.robust_list.head = &__main_pthread.robust_list.head;
   // Main thread ID is always 1.  It can't be 0 because musl assumes
   // tid is always non-zero.
   __main_pthread.tid = getpid();
   __main_pthread.locale = &libc.global_locale;
+  __main_pthread.next = __main_pthread.prev = &__main_pthread;
+
+  __main_pthread.tsd = (void **)__pthread_tsd_main;
```
Comment (kleisauke) [inline system/lib/pthread/library_pthread.c:849]: Since `->tsd` is now always initialized, I think we can get rid of this check:
https://github.com/emscripten-core/emscripten/blob/a47616119653aa89542dfac531722520fb16f0af/system/lib/libc/musl/src/thread/pthread_getspecific.c#L7-L10
(See previous discussion here: https://github.com/emscripten-core/emscripten/pull/14775/files#r678491019)

File: system/lib/pthread/library_pthread.c [hunk 11]
```diff
@@ -825,17 +825,26 @@ EMSCRIPTEN_KEEPALIVE void* _emscripten_main_thread_futex;
...
 // See system/lib/README.md for static constructor ordering.
 __attribute__((constructor(48)))
 void __emscripten_init_main_thread(void) {
   __emscripten_init_main_thread_js(&__main_pthread);
+  //printf("__emscripten_init_main_thread %p\n", __main_pthread);
+
   // The pthread struct has a field that points to itself - this is used as
   // a magic ID to detect whether the pthread_t structure is 'alive'.
   __main_pthread.self = &__main_pthread;
+  __main_pthread.detach_state = DT_JOINABLE;
```
Comment (kleisauke) [inline system/lib/pthread/library_pthread.c:840]: This should not be needed, as default thread state is already `DT_JOINABLE`.
https://github.com/emscripten-core/emscripten/blob/a47616119653aa89542dfac531722520fb16f0af/src/library_pthread.js#L804
(unless, I'm overlooking something)

Comment (sbc100) [reply system/lib/pthread/library_pthread.c:840]: That code in `library_pthread.js` is only run for pthread_create, not when the main/primordial thread is initialized.    This code here is *only* for main/primordial thread.

File: system/lib/pthread/library_pthread.c (hunk 10 above)
Comment (sbc100) [reply system/lib/pthread/library_pthread.c:849]: Yes, I think you are correct.

File: tools/system_libs.py [hunk 12]
```diff
@@ -732,7 +736,7 @@ def get_files(self):
     ignore = [
         'ipc', 'passwd', 'signal', 'sched', 'time', 'linux',
         'aio', 'exit', 'legacy', 'mq', 'setjmp', 'env',
-        'ldso'
+        'ldso', 'malloc'
```
[the remark this answers is not in the document]
Comment (sbc100) [reply tools/system_libs.py:739]: Interesting.  We already support various different allocators, and we never used the original musl one (although we also chose dlmalloc as the default).. so it should be easy enough to plug in and experiment with.

File: system/lib/pthread/library_pthread.c (hunk 11 above)
Comment (kleisauke) [reply system/lib/pthread/library_pthread.c:840]: Ah, I see. Looks like it was not previously initialized in musl's `__init_tp` (where we mirror this from), but now it is.
https://github.com/emscripten-core/emscripten/blob/3cc9f7ab560d634f94461848116d7b69c195cc58/system/lib/libc/musl/src/env/__init_tls.c#L21

Comment (sbc100) [reply system/lib/pthread/library_pthread.c:840]: > Ah, I see. Looks like it was not previously initialized in musl's `__init_tp` (where we mirror this from), but now it is.
> 
> https://github.com/emscripten-core/emscripten/blob/3cc9f7ab560d634f94461848116d7b69c195cc58/system/lib/libc/musl/src/env/__init_tls.c#L21

I was not aware of `__init_tp`.. I guess we could switch to that at some point in the future rather than duplicating it (in two places).

File: ChangeLog.md [hunk 13]
```diff
@@ -18,8 +18,13 @@ to browse the changes between the tags.
 
 See docs/process.md for more on how version tagging works.
 
-2.0.35
+3.0.0
 ------
+- The version of musl libc used by emscripten was upgraded from v1.1.15 to
+  v1.2.2.  There could be some minor size regressions (or gains) due to changes
+  in upstream musl code be we don't expect anything major.  Since this fairly
```
Comment (kripken) [inline ChangeLog.md:25]:
```suggestion
  in upstream musl code but we don't expect anything major.  Since this a fairly large
```

File: system/lib/libc/musl/arch/emscripten/bits/alltypes.h [hunk 14]
```diff
@@ -85,9 +91,9 @@ typedef long suseconds_t;
 #if defined(__NEED_pthread_attr_t) && !defined(__DEFINED_pthread_attr_t)
 typedef struct {
     union {
-        int __i[10];
-        volatile int __vi[10];
-        unsigned __s[10];
+        int __i[9];
+        volatile int __vi[9];
+        unsigned __s[9];
     } __u;
```
Comment (kripken) [inline system/lib/libc/musl/arch/emscripten/bits/alltypes.h:97]: I'm not sure how to review this change. Why did 10 become 9?

File: tests/test_other.py [hunk 15]
```diff
@@ -1667,8 +1667,8 @@ def test_dylink_pthread_static_data(self):
       emcc_args=[
         '-pthread', '-Wno-experimental',
         '-s', 'PROXY_TO_PTHREAD',
-        '-s', 'EXIT_RUNTIME=1',
-        '-s', 'MAIN_MODULE=1',
+        '-s', 'EXIT_RUNTIME',
+        '-s', 'MAIN_MODULE=2',
```
Comment (kripken) [inline tests/test_other.py:1671]: Looks like a reasonable change, just checking that I'm not missing anything - it's not strictly related to the PR?

File: tests/test_other.py [hunk 16]
```diff
@@ -4880,9 +4880,9 @@ def test_bad_locale(self):
 ''')
     self.run_process([EMXX, 'src.cpp'])
 
-    self.assertContained('locale set to C: C;C;C;C;C;C',
+    self.assertContained('locale set to C: C',
                          self.run_js('a.out.js', args=['C']))
-    self.assertContained('locale set to waka: waka;waka;waka;waka;waka;waka',
+    self.assertContained('locale set to waka: waka',
```
Comment (kripken) [inline tests/test_other.py:4885]: why did the output change here?

File: system/lib/pthread/pthread_create.c [hunk 17]
```diff
@@ -68,7 +68,35 @@ static long dummy_getpid() {
 }
 weak_alias(dummy_getpid, __syscall_getpid);
 
-int __pthread_create(pthread_t *restrict res, const pthread_attr_t *restrict attrp, void *(*entry)(void *), void *restrict arg) {
+static int tl_lock_count;
```
Comment (kripken) [inline system/lib/pthread/pthread_create.c:71]: Is all this logic copied from a musl file? If so which?

File: system/lib/pthread/library_pthread.c [hunk 18]
```diff
@@ -825,17 +825,26 @@ EMSCRIPTEN_KEEPALIVE void* _emscripten_main_thread_futex;
...
 void __emscripten_init_main_thread_js(void* tb);
 
+static void *dummy_tsd[1] = { 0 };
+weak_alias(dummy_tsd, __pthread_tsd_main);
+
 // See system/lib/README.md for static constructor ordering.
 __attribute__((constructor(48)))
 void __emscripten_init_main_thread(void) {
   __emscripten_init_main_thread_js(&__main_pthread);
+  //printf("__emscripten_init_main_thread %p\n", __main_pthread);
```
Comment (kripken) [inline system/lib/pthread/library_pthread.c:835]:
```suggestion
```

File: system/lib/libc/musl/include/threads.h [hunk 19]
```diff
@@ -20,6 +20,11 @@ typedef void (*tss_dtor_t)(void *);
 #define __NEED_cnd_t
 #define __NEED_mtx_t
 
+// XXX Emscripten: Fix musl libc issue, where the above defines are wrong, but instead should be the following: (https://github.com/emscripten-core/emscripten/issues/5343)
```
Comment (kripken) [inline system/lib/libc/musl/include/threads.h:23]: That issue is closed?

File: system/lib/libc/musl/src/locale/newlocale.c [hunk 20]
```diff
@@ -1,49 +1,73 @@
...
 
 	if ((loc = malloc(sizeof *loc))) *loc = tmp;
```
Comment (kripken) [inline system/lib/libc/musl/src/locale/newlocale.c:60]: Two lines above this (I can't open a comment there as there is a resolved one already) there is the same line but in a comment. Do we want the commented line, or the uncommented one?

File: system/lib/libc/musl/src/stat/fstatat.c [hunk 21]
```diff
@@ -1,10 +1,160 @@
...
+
+int fstatat(int fd, const char *restrict path, struct stat *restrict st, int flag)
 {
-	return syscall(SYS_fstatat, fd, path, buf, flag);
+	int ret;
+#ifdef __EMSCRIPTEN__
+	if (flag==AT_EMPTY_PATH && fd>=0 && !*path)
+		ret = __syscall(SYS_fstat, fd, st);
+	else if ((fd == AT_FDCWD || *path=='/') && !flag)
+		ret = __syscall(SYS_stat, path, st);
```
Comment (kripken) [inline system/lib/libc/musl/src/stat/fstatat.c:145]: This logic is copied from above, I think? Perhaps add a comment for that.

File: system/lib/libc/musl/src/stat/fstatat.c [hunk 22]
```diff
@@ -1,10 +1,160 @@
...
+#include <fcntl.h>
+#include <errno.h>
+#include <stdint.h>
+#include <sys/sysmacros.h>
 #include "syscall.h"
-#include "libc.h"
+#include "kstat.h"
 
-int fstatat(int fd, const char *restrict path, struct stat *restrict buf, int flag)
+/* XXX Emscripten: we ensure kstat == stat so we can simply make the syscall
```
Comment (kripken) [inline system/lib/libc/musl/src/stat/fstatat.c:11]: Can we add a static assert that the struct sizes match at least? Might avoid some future bugs.

File: system/lib/libc/musl/src/thread/pthread_getspecific.c [hunk 23]
```diff
@@ -4,10 +4,6 @@
 static void *__pthread_getspecific(pthread_key_t k)
 {
 	struct pthread *self = __pthread_self();
-	// XXX EMSCRIPTEN: self->tsd can be NULL in the case of the
-	// main thread where pthread_key_create is not called.
-	// See __pthread_key_create for where it get assigned.
-	if (!self->tsd) return NULL;
```
Comment (kripken) [inline system/lib/libc/musl/src/thread/pthread_getspecific.c:10]: why this change?

**Review by kripken — COMMENTED** [review]
Great to see this almost ready to land! Last comments from me.

File: system/lib/libc/musl/arch/emscripten/bits/alltypes.h (hunk 14 above)
Comment (sbc100) [reply system/lib/libc/musl/arch/emscripten/bits/alltypes.h:97]: This comes from upstream.  They unified these definitions here:

https://github.com/emscripten-core/musl/commit/2d3083e746cf314c16048875e919305b1018bf8b

See the new common definition in `alltypes.h.in`.

It does looks like we have an existing bug where these are too small when building with MEMORY64, but that is a separate/unrealted issue.

File: tests/test_other.py (hunk 15 above)
Comment (sbc100) [reply tests/test_other.py]: Reverted.

File: tests/test_other.py (hunk 16 above)
Comment (sbc100) [reply tests/test_other.py:4885]: I can't remember exactly but I think this was either bug in the existing version of musl or some detail that is allowed to vary.

Certainly if you run this under glibd you don't get the repeated field:

```
$ clang test.c
$ ./a.out 
locale set to C: C
```

So I think this is a bugfix, or at least harmless.

File: system/lib/pthread/pthread_create.c (hunk 17 above)
Comment (sbc100) [reply system/lib/pthread/pthread_create.c:71]: Yes, there is already comment at the top saying ` // See musl's pthread_create.c`.  Hopefully we can do a better job of code sharing as a followup.

File: system/lib/libc/musl/src/stat/fstatat.c (hunk 22 above)
Comment (sbc100) [reply system/lib/libc/musl/src/stat/fstatat.c]: Updated the comment.  We literally define one to be other.

File: system/lib/libc/musl/src/thread/pthread_getspecific.c (hunk 23 above)
Comment (sbc100) [reply system/lib/libc/musl/src/thread/pthread_getspecific.c:10]: As @kleisauke pointed out its no longer true that self->tsd can be NULL since we now initialize it on startup of both the main thread and any secondary threads.

Comment (sbc100) [reply system/lib/libc/musl/src/thread/pthread_getspecific.c:10]: I could perhaps split this change out..

Comment (kripken) [reply system/lib/libc/musl/src/thread/pthread_getspecific.c:10]: I see, thanks. sgtm as it is.

**Review by kripken — APPROVED** [review]
lgtm, although my understanding of the pthread bits is not perfect. @kleisauke do those look good to you?

File: system/lib/pthread/pthread_create.c [hunk 24]
```diff
@@ -111,8 +141,23 @@ int __pthread_create(pthread_t *restrict res, const pthread_attr_t *restrict att
...
+    return rtn;
+
+  __tl_lock();
+
+  new->next = self->next;
+  new->prev = self;
+  new->next->prev = new;
+  new->prev->next = new;
+
+  __tl_unlock();
```
Comment (kleisauke) [inline system/lib/pthread/pthread_create.c:156]: This was introduced in musl commit [`8f11e61`](https://git.musl-libc.org/cgit/musl/commit/?id=8f11e6127fe93093f81a52b15bb1537edc3fc8af) (and [`ba74a42`](https://git.musl-libc.org/cgit/musl/commit/?id=ba74a42cee90c9a4425188a021b6ad8ba80b9468) for `pthread_key_delete`), but I think using a global thread list is not necessary in WASM.

I had a go to avoid this newly added logic, see for example commit https://github.com/emscripten-core/emscripten/pull/13007/commits/bbc1faab026280ffb574bb57b9a54e34bfe402c5, WDYT?

**Comment by kleisauke** [comment]
The test failure in `browser.test_html5_webgl_api` can be resolved with:
```diff
--- a/src/library_pthread.js
+++ b/src/library_pthread.js
@@ -684,7 +684,7 @@ var LibraryPThread = {
     // Deduce which WebGL canvases (HTMLCanvasElements or OffscreenCanvases) should be passed over to the
     // Worker that hosts the spawned pthread.
     // Comma-delimited list of CSS selectors that must identify canvases by IDs: "#canvas1, #canvas2, ..."
-    var transferredCanvasNames = attr ? {{{ makeGetValue('attr', 40, POINTER_TYPE) }}} : 0;
+    var transferredCanvasNames = attr ? {{{ makeGetValue('attr', 36/*_a_transferredcanvases*/, POINTER_TYPE) }}} : 0;
 #if OFFSCREENCANVASES_TO_PTHREAD
     // Proxied canvases string pointer -1 is used as a special token to fetch
     // whatever canvases were passed to build in -s
```

**Review by kleisauke — APPROVED** [review]
pthread bits looks good to me, I left a small comment about avoiding the global thread list, but that can also be incorporated later.

(Note that I have pending changes that may land after this PR, see comment https://github.com/emscripten-core/emscripten/pull/13007#issuecomment-804826213 for an overview)

File: system/lib/pthread/pthread_create.c (hunk 24 above)
Comment (kleisauke) [reply system/lib/pthread/pthread_create.c:156]: Oh, it seems that avoiding the global thread list also fixes the remaining `browser.test_pthread_gcc_atomic*` failures on Chrome.

FWIW, commit 4dfe1d60cfb4aeea41e3ad14bc0b7bfd7f47b085 is easier to test with (since bbc1faab026280ffb574bb57b9a54e34bfe402c5 was based on PR #13007).

Comment (sbc100) [reply system/lib/pthread/pthread_create.c:156]: The linked list of threads is used internally within musl, so think its good to mirror that if we can.   For example, in __pthread_key_delete.

Is there any reason not to maintain it?  (assuming we can fix the remainging failures?)  Is there any reason you think maintaining the list list is less necessary in wasm?

Comment (kleisauke) [reply system/lib/pthread/pthread_create.c:156]: Other than this reduces code size and complexity, there was no particular reason for including this commit. It didn't cause any regressions either, but you're right, we probably shouldn't deviate too much from musl.

Hopefully we can fix the remaining test failures so this PR can land (I do wonder why that commit seemed to fix those failures).

**Comment by kripken** [comment]
Excellent! Thanks for all the work here @sbc100 & @kleisauke !
